### DISCUSSION TITLE
Using forwardRef under a Popover.Trigger

### DISCUSSION THREAD

**Discussion by samselikoff** [radix-ui/primitives#1940] [Help] [answered]
Hi!

I'm having trouble using a custom component with forwardRef under a Popover.Trigger. First time doing this with this lib so would appreciate if anybody can point me in the right direction!

Here's the repo:

https://github.com/samselikoff/2023-02-07-radix-popover-child-component

And here's the Stackblitz:

https://stackblitz.com/github/samselikoff/2023-02-07-radix-popover-child-component?file=pages%2Findex.tsx

If you comment out my custom `<Button>` you'll see everything works. For some reason I can't get the ref forwarding to take!

**Comment by joaom00** [comment, marked as answer]
Hi @samselikoff,

You should forward all props too
```jsx
const Button = forwardRef(function Button(
  { children, ...props }: { children: ReactNode },
  ref
) {

  return (
    <button {...props} className="foo" ref={ref}>
      {children}
    </button>
  );
});
```

**Reply by benoitgrelard** [reply]
Yep, as @joaom00 pointed out, you need to make sure your button accepts handlers that will be added to it by `Popover.Trigger` to open the trigger using pointer/keyboard, but also accept data attributes, aria attributes, etc. The easiest way to do that without having to specifically list out these is to spread all the props coming in, which makes sense anyway for leaf components like buttons in a design system.

**Reply by samselikoff** [reply]
Makes sense! Thanks both for the answer!

Follow-up, is it possible to get types for the `...props`? I'm using `props.onClick` in a Framer Motion hook and feel like that's kinda reaching into private API 🤔

**Reply by benoitgrelard** [reply]
If you mean for your case above, the type of props would be:
```ts
React.ComponentProps<'button'>
```

**Reply by samselikoff** [reply]
I actually mean the specific forwarded props from the `Popover.Trigger`. It would be nice to know which props are forwarded – not sure if that's possible?

Also for the props in the above case if I wanted all of them, wouldn't it be `ComponentPropsWithoutRef<'button'>`, since I'm using forwardRef?

**Reply by benoitgrelard** [reply]
> I actually mean the specific forwarded props from the Popover.Trigger. It would be nice to know which props are forwarded – not sure if that's possible?

Ah I see sorry! No that's not possible, in an ideal world, TS would moan if you `asChild` and the direct child doesn't have the necessary stuff, or the right signature, but as far as we know that isn't possible with TS at the moment.

> Also for the props in the above case if I wanted all of them, wouldn't it be ComponentPropsWithoutRef<'button'>, since I'm using forwardRef?

Yep, that's correct. That's how we extend all our primitives internally too.

**Comment by samselikoff** [comment]
Makes sense. Thanks for all the info + awesome work on this lib! 🙏

**Comment by alainkaiser** [comment]
Thanks for that! Saved my day 👍🏼
